From cd931fd0f33705b7b9ac7bdeb0907dfa17a3ee0e Mon Sep 17 00:00:00 2001
From: Ondrej Jirman <megous@megous.com>
Date: Mon, 30 Nov 2020 06:12:29 +0100
Subject: [PATCH 259/351] usb: typec: anx7688: Kill the load upon receiving
 hard reset

FW likes to send hard resets for bullshit reasons, so only kill the load
if the power was negotiated via PD messages.

Signed-off-by: Ondrej Jirman <megous@megous.com>
---
 drivers/usb/typec/anx7688.c | 19 ++++++++++++++++++-
 1 file changed, 18 insertions(+), 1 deletion(-)

diff --git a/drivers/usb/typec/anx7688.c b/drivers/usb/typec/anx7688.c
index 39c817e0235f..0db7d93c41f7 100644
--- a/drivers/usb/typec/anx7688.c
+++ b/drivers/usb/typec/anx7688.c
@@ -879,7 +879,24 @@ static int anx7688_handle_pd_message(struct anx7688* anx7688,
 		break;
 
 	case ANX7688_OCM_MSG_HARD_RST:
-		dev_info(anx7688->dev, "received HARD_RST\n");
+		if (anx7688->pd_capable) {
+			dev_info(anx7688->dev, "received HARD_RST\n");
+
+			// stop drawing power from VBUS
+			psy_val.intval = 0;
+			dev_dbg(dev, "disabling vbus_in power path\n");
+			ret = power_supply_set_property(anx7688->vbus_in_supply,
+							POWER_SUPPLY_PROP_ONLINE,
+							&psy_val);
+			if (ret)
+				dev_err(anx7688->dev, "failed to offline vbus_in\n");
+
+			// wait till the dust settles
+			mod_timer(&anx7688->nopd_timer, jiffies + msecs_to_jiffies(3000));
+		} else {
+			dev_dbg(anx7688->dev, "received HARD_RST, idiot firmware is bored\n");
+		}
+
 		break;
 
 	case ANX7688_OCM_MSG_RESTART:
-- 
2.34.0

